### PR TITLE
BIP draft: Raw() as subscript for descriptors

### PR DESCRIPTION
**Pull request by furszy**
Allowing arbitrary hex data to be wrapped in `sh()`, `wsh()`, or even within the `TREE`
argument of a `tr(KEY, TREE)` descriptor enables the representation of currently
inexpressible information.

Specifically, the absence of this feature limits the representation of non-standard redeem
and witness scripts. This occurs because they can currently only be represented as top-level
`raw(HEX)` descriptors, which retain only the output script information and lack the ability
to preserve the actual script.

This work stemmed from discussions with @sipa and @achow101 about the legacy
wallet migration path, particularly while investigating migrated descriptors that lacked
information about the script they were supposedly migrated from.

Additionally, issue [#24114](https://github.com/bitcoin/bitcoin/issues/24114), specifically https://github.com/bitcoin/bitcoin/issues/24114#issuecomment-1127978154, provides
historical context and highlights other valid use cases enabled by this behavior change.

### CODE REVIEW

File: bip-0391.mediawiki [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+<tt>raw()</tt> encapsulates a raw hex script. This BIP allows <tt>raw()</tt> to be used in
+the context of other descriptors.
+
+==Copyright==
+
+This BIP is licensed under the BSD 2-clause license.
+
+==Motivation==
+
+To ensure no information is lost during conversion to descriptors, it’s useful to wrap
```
[the remark this answers is not in the document]
Comment (furszy) [reply bip-0391.mediawiki]: absolutely, thanks!

File: bip-0391.mediawiki [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+Additionally, as noted [https://github.com/bitcoin/bitcoin/issues/24114#issuecomment-1127978154 here],
+there are other useful scenarios for this feature. For example, it allows representing
+in a descriptor that we lack full knowledge of all solvability paths but can still solve
+the output, such as a taproot tree where we know only one of its paths, or even also
+participating in signing a script without knowing all participants' keys, relying solely
+on the script structure.
+
+==Specification==
+
+Allow raw(HEX) to be accepted within the context of `sh()`, `wsh()` and `tr()` descriptors.
```
Comment (murchandamus) [inline bip-0391.mediawiki:44]: I think you’re mixing up MediaWiki and Markdown. If you want to use mediawiki, you would use `<tt>sh()</tt>`. It’s been permitted to submit BIPs in Markdown for a while though, if you prefer that.

```suggestion
Allow raw(HEX) to be accepted within the context of <tt>sh()</tt>, <tt>wsh()</tt> and <tt>tr()</tt> descriptors.
```

Comment (furszy) [reply bip-0391.mediawiki:44]: > It’s been permitted to submit BIPs in Markdown for a while though, if you prefer that.

Ok yes, that's better. Will inline the doc to use the `` markdown syntax.

File: bip-0391.mediawiki [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+and witness scripts. This occurs because they can currently only be represented as top-level
+`raw(HEX)` descriptors, which retain only the output script information and lack the ability
+to preserve the actual script.
+
+Additionally, as noted [https://github.com/bitcoin/bitcoin/issues/24114#issuecomment-1127978154 here],
+there are other useful scenarios for this feature. For example, it allows representing
+in a descriptor that we lack full knowledge of all solvability paths but can still solve
+the output, such as a taproot tree where we know only one of its paths, or even also
+participating in signing a script without knowing all participants' keys, relying solely
+on the script structure.
```
Comment (murchandamus) [inline bip-0391.mediawiki:41]: This paragraph is hard to parse, perhaps you could split the second sentence into multiple sentences.

File: bip-0391.mediawiki [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+==Specification==
+
+### For `sh()` and `wsh()` descriptors:
+`raw(HEX)` must represent the arbitrary script data within the provided context.
+This for example means that a P2SH output script (OP_HASH160 <hash160(HEX_script)> OP_EQUAL)
+must be created from the provided hex data when a `sh(raw(HEX))` descriptor is provided.
+
+Parallely, a P2WSH output script (OP_0 <hash160(HEX_script)>) must be created from the provided
```
Comment (murchandamus) [inline bip-0391.mediawiki:50]:
```suggestion
Parallelly, a P2WSH output script (OP_0 <hash160(HEX_script)>) must be created from the provided
```

File: bip-0391.mediawiki [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+Invalid descriptors
+
+* Non-hex script: `sh(raw(asdf))`
+* Non 32-bit hash in `rawnode`: `tr(a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd, rawnode(kjke))`
+* Non-hex in `rawnode`: `tr(a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd, rawnode(<complete me with a 32-bit non-hex hash>))`
+* `raw` in the key path `tr`: `tr(raw(asdf), pk(key))`
+
+==Backwards Compatibility==
+
+`raw()` as a subscript descriptors use the format and general operation specified in [[bip-0380.mediawiki|380]].
```
Comment (murchandamus) [inline bip-0391.mediawiki:93]:
```suggestion
`raw()` as subscript descriptors use the format and general operation specified in [[bip-0380.mediawiki|380]].
```

File: bip-0391.mediawiki [hunk 6]
```diff
@@ -0,0 +1,98 @@
+<pre>
+  BIP: ???
+  Layer: Applications
+  Title: raw() as subscript in Output Script Descriptors
+  Author: Pieter Wuille pieter@wuille.net
+          Matias Furszyfer mfurszy@protonmail.com
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-???
+  Status: Draft
+  Type: Informational
```
Comment (murchandamus) [inline bip-0391.mediawiki:10]:
```suggestion
  Type: Standards Track
```

File: bip-0391.mediawiki [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+  Created: 2024-12-16
```
Comment (murchandamus) [inline bip-0391.mediawiki:11]: The Created Header stores the date a BIP was assigned a number:

```suggestion
  Created: ?
```

File: bip-0391.mediawiki [hunk 8]
```diff
@@ -0,0 +1,98 @@
+<pre>
+  BIP: ???
+  Layer: Applications
+  Title: raw() as subscript in Output Script Descriptors
+  Author: Pieter Wuille pieter@wuille.net
+          Matias Furszyfer mfurszy@protonmail.com
```
Comment (murchandamus) [inline bip-0391.mediawiki:6]:
```suggestion
  Author: Pieter Wuille <pieter@wuille.net>
          Matias Furszyfer <mfurszy@protonmail.com>
```

**Review by murchandamus — COMMENTED** [review]
Could maybe use a bit more rationale. Has this already been submitted to the mailing list?

**Comment by furszy** [comment]
> Has this already been submitted to the mailing list?

Not yet. Will do next year.

File: bip-raw-as-subscript.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+Specifically, the absence of this feature limits the representation of non-standard redeem
+and witness scripts. This occurs because they can currently only be represented as top-level
+`raw(HEX)` descriptors, which retain only the output script information and lack the ability
+to preserve the actual script.
+
+Additionally, as noted [here](https://github.com/bitcoin/bitcoin/issues/24114#issuecomment-1127978154),
+there are other useful scenarios for this feature. For example, it allows representing
+in a descriptor that we lack complete knowledge of all solvability paths but can still
+solve the output. This includes cases like a taproot tree where we know only one of its
```
Comment (ismaelsadeeq) [inline bip-raw-as-subscript.md:38]: This is a bit unclear for me, can you clarify what do you mean by can still solve the output?

File: bip-raw-as-subscript.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+#### `rawnode(HEX)`:
+Indicating a tree node with specified 32-bit hash, but no specified subtree.
+This can serve as either a tree branch or the root of the Merkle tree.
+
+#### `raw(HEX)`:
+Defines a tree leaf containing the specified script in hex.
+Note: The leaf version must be internally fixed at the existing `0xC0` to prevent introducing
+unsupported or undefined functionality.
+If a different version is required for any use case, a new BIP could introduce `raw(HEX, VERSION)`
+in the future.
```
Comment (ismaelsadeeq) [inline bip-raw-as-subscript.md:64]: I think we should just support `raw(HEX, VERSION)` and be forward compatible which goes in line with what was suggested in the https://github.com/bitcoin/bitcoin/issues/24114

**Comment by achow101** [comment]
Concept ACK I guess

**Comment by murchandamus** [comment]
Hey @furszy, has this been presented to the mailing list? Did you see @ismaelsadeeq’s review?

**Comment by jonatack** [comment]
Hi @furszy, are you still working on this BIP draft?

**Comment by furszy** [comment]
Sorry for the delay. I started writing some updates a while ago but haven’t had enough time to finish them.
Unfortunately, I won’t have time over the next two months either. The idea was also to check https://github.com/bitcoin/bitcoin/pull/30243 so we could align some concepts, but that’s still buried in my work queue as well.
No problem at all if you want to close it so someone else can take over. I can always pick it up again later.

**Review by murchandamus — COMMENTED** [review]
Given that there hasn’t been too much activity in this PR so far, I’m gonna close it for now. Please feel free to ask to have it reopened in the future or to open another PR after it has been discussed on the mailing list and is gaining a bit of momentum.

**Comment by murchandamus** [comment]
Given that there hasn’t been too much activity in this PR and, I’m gonna close it for now. Please feel free to ask to have it reopened in the future or to open another PR after it has been discussed on the mailing list and is gaining a bit of momentum.
